api/v1alpha1/instancegroup_types.go

```diff
@@ -47,8 +47,9 @@ const (
 	ReconcileModified  ReconcileState = "ReconcileModified"
 
 	// End States
-	ReconcileReady ReconcileState = "Ready"
-	ReconcileErr   ReconcileState = "Error"
+	ReconcileLocked ReconcileState = "Locked"
```

Comment:
Not sure "Locked" should be a reconcile state - the intention is to lock upgrades, not reconciles.

Reply:
I'm missing something here, what's an example of a reconcile that doesn't result in an update/upgrade?

from my understanding; each change to the `CR` will trigger the state to go to `Init` eg.

```
2021-08-05T22:22:09.340+1000    INFO    v1alpha1        state transition occured        {"instancegroup": "instance-manager/platform-us-west-2b", "state": "Init", "previousState": "Locked"}
```

and then flow following:

```
func HandleReconcileRequest(d CloudDeployer) error 
```

I've used the reconcile state to mark that:
1. The `IG` should show it's `Locked`
2. The AWS resources should not be modified by preventing entering the `ReconcileInitUpdate`, `ReconcileInitUpgrade` 

remaining questions:
1. Is there a difference between a reconcile and an upgrade/update?
2. Is my understanding of the reconcile state flow correct?
3. where does it miss the design we've discussed so far?

thanks for your help :)

Reply:
For showing something on kubectl we can also embed `locked: true/false` in status and display it as a new column if we wanted - we don't have to use reconcile state, but I think from display perspective it makes sense.

Your questions:
1. Reconcile-update is for non node rotating changes, e.g. add an EC2 tag. or basically any change that is not on the launch template / launch config. reconcile-upgrade would be changes that cause a drift in the launch config / template.
2. Yes, I think so, the interface shows the flow pretty clearly - if you put `Lock` at first it means you are blocking other flows from executing.
3. I think the original intention was to only block node-rotation. However with this design we are blocking updates to ASG which do not cause node rotations - while that also might be a nice feature we should consider the use cases around it, and also reconsider the annotation name to simply be `"instancemgr.keikoproj.io/locked"` in this case.

Reply:
Whats the flow you would expect for a locked IG?

Should we bypass the upgrade flow?
`Init` -> `InitUpgrade` -> `Ready` 

by adding a condition on the `UpgradeNodes()` call: 

```
if d.GetState() == v1alpha.ReconcileInitUpgrade && !d.Locked() {
  err = d.UpgradeNodes()
```

Should we

`Init` -> `InitUpgrade` -> `ReconcileModifying` -> `ReconcileModified` -> `Ready` 

and shortcut the flow in `UpgradeNodes()`:

```
func (ctx *EksInstanceGroupContext) UpgradeNodes() error {
  var (
... snip ...
  )

  if ctx.Configuration.InstanceGroup.Locked() {
    return nil
  }
```

or something else?

Reply:
What if we just short circuit from the interface? would that work?
e.g. 
```golang
if d.Locked() {
  d.SetState(v1alpha.ReconcileLocked)
  return nil
}
```